File: ruby/binary-search/binary_search.rb
class BinarySearch
  def search_for(target)
    low = 0
    high = list.length
    mid = low + (high - low) / 2

    while low < high
      case target <=> list[mid]
      when -1
        high = mid
        mid = low + (high - low) / 2
      when 0
        return mid
      when 1
        low = mid + 1
        mid = low + (high - low) / 2
      end
    end
  end

  private

  attr_accessor :list

  def initialize(list)
    @list = list
  end
end

File: clojure/binary-search/src/binary_search.clj
(ns binary-search)

(defn middle [coll]
  (let [size (count coll)]
    (int (/ size 2))))

(defn search-for [n coll]
  (loop [low 0 high (count coll)]
    (if (<= high low)
      (throw (Exception. "not found"))
      (let [mid (int (+ low (/ (- high low) 2)))]
        (case (compare n (nth coll mid))
          -1 (recur low mid)
          0  mid
          1  (recur (inc mid) high))))))
